modules/login/php/ajaxlogin.class.inc

```diff
@@ -0,0 +1,405 @@
+        $reCAPTCHAPublic    = $config->getSetting('reCAPTCHAPublic');
+        // reCAPTCHA if both private and public keys are set
+        if ($reCAPTCHAPrivate && $reCAPTCHAPublic) {
+            $requestAccountData['captcha'] = $reCAPTCHAPublic; // captcha key.
+        }
+        $requestAccountData['site'] = \Utility::getSiteList();
+        $values['requestAccount']   = $requestAccountData;
+
+        return $values;
+    }
+
+    /**
+     * Reset password.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function resetPassword(array $values) : array
+    {
+        // Always claim success to phishers.
+        $values['success'] = 'Password reset. '
+            . 'You should receive an email within a few minutes.';
+
+        // Create the user object.
+        $username = $values['username'];
+        $user     =& \User::singleton($username);
+
+        // Check if valid user.
+        $email = '';
+        try {
+            $email = $user->getData('Email');
+            if (empty($email) || empty($user)) {
+                $values['error'] = 'invalid';
+                return $values;
+            }
+        } catch (\Exception $exception) {
+            $values['error'] = 'invalid';
+        }
+
+        try {
+            // Check if valid email.
+            if ($user->isEmailValid()) {
+                $config = \NDB_Config::singleton();
+                // Generate a new password.
+                $password = \User::newPassword();
+                // Reset the password in the database and
+                // expire password so user must change it upon login.
+                $user->updatePassword($password, '1999-01-01');
+                // Send the user an email.
+                $msg_data['study']    = $config->getSetting('title');
+                $msg_data['url']      = $config->getSetting('url');
+                $msg_data['realname'] = $user->getData('Real_name');
+                $msg_data['password'] = $password;
+                \Email::send($email, 'lost_password.tpl', $msg_data);
+            } else {
+                $values['error'] = 'invalid';
+                error_log(
+                    $_SERVER['REMOTE_ADDR']
+                    . '(' . $_SERVER['HTTP_X_FORWARDED_FOR'] . ')'
+                    . " attempted to reset password for $username"
+                );
+            }
+        } catch(\Exception $e) {
+            error_log('Unhandled exception in reset_password.');
+            $values['error'] = 'invalid';
+        }
+        return $values;
+    }
+
+    /**
+     * Request account.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     * @throws \DatabaseException
+     */
+    function requestAccount(array $values) : array
+    {
+        // Verify reCAPTCHA
+        if (isset($values['g-recaptcha-response']) && isset($reCAPTCHAPrivate)) {
```

Comment:
`$reCAPTCHAPrivate` doesn't appear to be defined in this scope.

Reply:
I'm going to make a separate PR about this but reCAPTCHAPrivate isn't used anymore by Google. Since they have their own authentication for domains with only needing a public key.

Reply:
So none of this code works anyway? Might as well delete it now then.

Reply:
@johnsaigle I tested the frontend with adding a public key of my own and the reCaptcha does display on LORIS on this PR. I haven't debugged the backend but I assume it never has worked and nobody has ever used reCaptcha. Since the Content-Security was never allowing reCaptcha in the first place because iframe-src would have broke it and I had to add that rule.

Reply:
Yeah I'm not surprised it never worked or hasn't for a long time. Definitely was the case for me but I never looked into it closely.

So you're saying that this code is more old i.e. not how Google does things anymore BUT it does work now, right?

Reply:
@johnsaigle the frontend now shows reCaptcha but I didn't debug through the backend to see if it matters or not.

Reply:
Yeah it would be a good idea to test if the reCaptcha actually works rather than if it just shows up.